Escrow-JS Example (#1271)
near/docs: docs/3.tutorials/examples/escrow.md

---
id: escrow
title: Escrow Contract
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import {CodeTabs, Language, Github} from "@site/components/codetabs"

> This example provides a simple example for an escrow contract, with a timeout callable function that concludes the transaction if the buyer has not provided an approval within a given timeframe. 

Reviewer: ```suggestion
This example provides a simple implementation of an escrow contract, with a timeout callable function that concludes the transaction if the buyer has not provided approval within a given timeframe. 
```

The code also includes an example of how contracts can pass data to each other through cross contract calls, and how the caller contract can handle the result.

:::info
This example does not have a frontend
:::

---

## Starting with the Project

> You have two options to start using the project. The first and recommended is to open the project through Gitpod, which will open a web-based interactive environment with all dependencies installed. The second option is to clone the repository locally, for which you will need to install all the [Prerequisites](../../2.develop/prerequisites.md).

Reviewer: ```suggestion
You have two options to start using the project:
- The first and recommended is to open the project through Gitpod, which will open a web-based interactive environment with all dependencies installed.
- The second option is to clone the repository locally, for which you will need to install all the [Prerequisites](../../2.develop/prerequisites.md).
```

  | Gitpod                                                                                                                                                                                           | Clone Locally                                                                 |
  | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------------------------------------- |
  | <a href="https://gitpod.io/#https://github.com/near-examples/escrow-js"><img src="https://gitpod.io/button/open-in-gitpod.svg" alt="Open in Gitpod" /></a> | 🌐 `https://github.com/near-examples/escrow-js.git` |

  </TabItem>
</Tabs>

---

### Interacting With the Contract
Since this example does not have a frontend, we will interact with it using the [NEAR CLI](../../4.tools/cli.md) or writing tests using [`workspaces-js`](../../2.develop/testing/integration.md).

Check the README.md. Briefly, you will need to:

#### 1. Install Dependencies
To install dependencies, run:

```bash
npm i
```

#### 1. Build the Contracts

> You can build the contract (creating the `.wasm` file that is ultimately deployed to the network) by running:

Reviewer: ```suggestion
You can build the contracts (creating the `.wasm` files that will be deployed to the network) by running:
```

```bash
near view <your-assets-testnet-account-id> get_account_assets '{"account_id": "<your-account-id>"}'
```

You may also check the NEAR balance of the seller account, making sure they have not received the payment yet:

```bash
near state <seller-account-id>
```

#### 5. Approve the Purchase

To approve the purchase, run:

```bash
near call <your-escrow-testnet-account-id> approve_purchase '{}' --accountId <your-account-id>
```

Check the NEAR balance of the seller account again, and you should see that they have received the payment.

---

### Contract

The contract exposes methods to purchase assets in escrow and approve or cancel the purchase by the buyer. Additionally, the contract exposes a method to check the creation time of the purchase, and if the purchase has been created a day or longer ago, the contract will approve the purchase without the buyer's input. 

<CodeTabs>
<Language value="🌐 JavaScript" language="js">
    <Github fname="contract.ts" 
            url="https://github.com/near-examples/escrow-js/blob/master/contracts/escrow.js"
            start="41" end="119" />
  </Language>
</CodeTabs>

---

## Testing

> When writing smart contracts it is very important to test all methods exhaustively. In this

Reviewer: ```suggestion
When writing smart contracts it is critical to test all methods exhaustively. In this
```
